ddht/v5_1/network.py

```diff
@@ -130,67 +121,161 @@ async def do_lookup(node_id: NodeID) -> None:
     )
 
 
-async def common_get_nodes_near(
-    network: NetworkProtocol, target: NodeID, *, max_nodes: int = 32,
-) -> Tuple[NodeID, ...]:
+async def common_explore_target(
+    network: NetworkProtocol,
+    target: NodeID,
+    *,
+    max_nodes: int = 32,
+    concurrency: int = 3,
+) -> Tuple[ENRAPI, ...]:
     """
     Lookup nodes in the target area of the routing table.
 
     Unlike `recursive_find_nodes` which is good at finding the very closest
     node, this function aims to find as many nodes that are as close to the
     `target` as possible.
     """
-    enrs_from_recursive_find = await network.recursive_find_nodes(target)
+    node_ids_to_query: List[NodeID]
+    node_ids_to_query = list(network.routing_table.iter_nodes_around(target))
 
-    for enr in enrs_from_recursive_find:
-        network.enr_db.set_enr(enr)
+    # Track all of the node ids we have seen.
+    seen_node_ids: Set[NodeID] = {network.local_node_id}
+    seen_node_ids.update(node_ids_to_query)
 
-    node_ids_from_routing_table = tuple(network.routing_table.iter_nodes_around(target))
-    node_ids_from_recursive_find = tuple(
-        enr.node_id
-        for enr in enrs_from_recursive_find
-        if enr.node_id != network.local_node_id
-    )
+    # Track what node ids we have already queried
+    queried_node_ids: Set[NodeID] = {network.local_node_id}
```

Comment:
nit: hard to keep track of all these ID sets, `seen`, `to_query`, `queried`, `in_flight` (i think that's all of the states?). i think the distinction between `seen` and `to_query` is least clear. might be worth documenting this lifecycle in the docstring.